lib/chef/resource/powershell_package_source.rb

```diff
@@ -62,105 +156,167 @@ class PowershellPackageSource < Chef::Resource
-        # TODO: Ensure package provider is installed?
-        if psrepository_cmdlet_appropriate?
-          if package_source_exists?
-            converge_if_changed :url, :trusted, :publish_location, :script_source_location, :script_publish_location do
-              update_cmd = build_ps_repository_command("Set", new_resource)
-              res = powershell_exec(update_cmd)
-              raise "Failed to update #{new_resource.source_name}: #{res.errors}" if res.error?
-            end
-          else
-            converge_by("register source: #{new_resource.source_name}") do
-              register_cmd = build_ps_repository_command("Register", new_resource)
-              res = powershell_exec(register_cmd)
-              raise "Failed to register #{new_resource.source_name}: #{res.errors}" if res.error?
-            end
+      # Notes:
+      # There are 2 objects we care about with this code. 1) The Package Provider which can be Nuget, PowerShellGet, Chocolatey, et al. 2) The PackageSource where the files we want access to live. The Package Provider gets us access to the Package Source.
+      # Per the Microsoft docs you can only have one provider for one source. Enter the PSRepository. It is a sub-type of Package Source.
+      # If you register a new PSRepository you get both a PSRepository object AND a Package Source object which are distinct. If you call "Get-PSRepository -Name 'PSGallery'" from powershell, notice that the Packageprovider is Nuget
+      # now go execute "Get-PackageSource -Name 'PSGallery'" and notice that the PackageProvider is PowerShellGet. If you set a new PSRepository without specifying a PackageProvider ("Register-PSRepository -Name 'foo' -source...") the command will create both
+      # a PackageSource and a PSRepository with different providers.
+
+      # Unregistering a PackageSource (unregister-packagesource -name 'foo') where that source is also a PSRepository also causes that object to delete as well. This makes sense as PSRepository is a sub-type of packagesource.
+      # All PSRepositories are PackageSources, and all PackageSources with Provider PowerShellGet are PSRepositories. They are 2 different views of the same object.
+
+      action :register, description: "Registers a PowerShell package source." do
+        package_details = get_package_source_details
+        output = package_details.result
+        if output == "PSRepository" || output == "PackageSource"
+          action_set
+        elsif new_resource.provider_name.downcase.strip == "powershellget"
+          converge_by("register source: #{new_resource.source_name}") do
+            register_cmd = build_ps_repository_command("Register", new_resource)
+            res = powershell_exec(register_cmd)
+            raise "Failed to register #{new_resource.source_name}: #{res.errors}" if res.error?
           end
         else
-          if package_source_exists?
```

Comment:
register also likely needs to have a check for package_source_exists and it needs idempotency checks to make sure it doesn't fire a second time if the source is already registered.